pkg/blobstore/grpcclients/cas_zstd_blob_access.go

```diff
@@ -0,0 +1,265 @@
+) blobstore.BlobAccess {
+	if compressionThresholdBytes == 0 {
+		compressionThresholdBytes = _defaultCompressionThresholdBytes
+	}
+
+	return &casWithZstdBlobAccess{
+		uncompressedBlobAccess:          NewCASBlobAccess(client, uuidGenerator, readChunkSize),
+		byteStreamClient:                bytestream.NewByteStreamClient(client),
+		contentAddressableStorageClient: remoteexecution.NewContentAddressableStorageClient(client),
+		capabilitiesClient:              remoteexecution.NewCapabilitiesClient(client),
+		uuidGenerator:                   uuidGenerator,
+		readChunkSize:                   readChunkSize,
+		compressionThresholdBytes:       compressionThresholdBytes,
+	}
+}
+
+type zstdByteStreamChunkReader struct {
+	client        bytestream.ByteStream_ReadClient
+	cancel        context.CancelFunc
+	zstdReader    io.ReadCloser
+	readChunkSize int
+	wg            sync.WaitGroup
+}
+
+// Read opens a pipe which allows us to process the compressed stream into the ZSTD
+// reader without blocking or keeping chunks in memory.
+//
+// Unlike the non-compressed version that can return gRPC chunks directly, compression
+// requires bridging two incompatible interfaces:
+// - gRPC pushes chunks to us via client.Recv().
+// - ZSTD expects to pull data from us via Read().
+//
+// We work around this by using a goroutine that receives gRPC chunks and writes them to a pipe,
+// while the ZSTD decoder reads from the other end of the pipe. This creates a streaming
+// pipeline: gRPC -> goroutine -> pipe -> ZSTD -> CAS.
+func (r *zstdByteStreamChunkReader) Read() ([]byte, error) {
+	if r.zstdReader == nil {
+		pr, pw := io.Pipe()
+
+		r.wg.Add(1)
+		go func() {
+			defer r.wg.Done()
+			defer pw.Close()
+			for {
+				chunk, err := r.client.Recv()
+				if err != nil {
+					if err != io.EOF {
+						pw.CloseWithError(err)
+					}
+					return
+				}
+				if _, writeErr := pw.Write(chunk.Data); writeErr != nil {
+					pw.CloseWithError(writeErr)
+					return
+				}
+			}
+		}()
+
+		var err error
+		r.zstdReader, err = util.NewZstdReadCloser(pr, zstd.WithDecoderConcurrency(1))
+		if err != nil {
+			pr.Close()
+			return nil, err
+		}
+	}
+
+	buf := make([]byte, r.readChunkSize)
+	n, err := r.zstdReader.Read(buf)
+	if n > 0 {
+		if err != nil && err != io.EOF {
+			err = nil
+		}
+		return buf[:n], err
+	}
+	return nil, err
+}
+
+func (r *zstdByteStreamChunkReader) Close() {
+	if r.zstdReader != nil {
+		r.zstdReader.Close()
+	}
+	r.cancel()
+
+	// Drain the gRPC stream.
+	for {
+		if _, err := r.client.Recv(); err != nil {
+			break
+		}
+	}
+	r.wg.Wait()
+}
+
+func (ba *casWithZstdBlobAccess) Get(ctx context.Context, digest digest.Digest) buffer.Buffer {
+	if digest.GetSizeBytes() < ba.compressionThresholdBytes {
+		return ba.uncompressedBlobAccess.Get(ctx, digest)
+	}
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	resourceName := digest.GetByteStreamReadPath(remoteexecution.Compressor_ZSTD)
+	client, err := ba.byteStreamClient.Read(
```

Comment:
This currently fails to do negotiation with the server to see whether compression is supported. It would be nice if the `GetCapabilities()` method of this backend kept an eye on the `SupportedCompressors` returned by the server and have `Get()` use `Put()` respect this. Furthermore, if `Get()` or `Put()` is called before any call to `GetCapabilities()` is made, it may call `GetCapabilities()` directly.